### ISSUE TITLE
Deploying the Docker image on Google Cloud Platform

### ISSUE THREAD

**Issue by mohanrao** [andrieslouw/imagesweserv#179]
Hi,

I am trying to host weserv on GCP.

Any references like using with docker and how to scale etc.

**Comment by mohanrao** [comment]
Original reason to host is google is blocked by the robots.txt . Any ways to support crawlers ?

**Comment by andrieslouw** [comment]
Hi there, sorry for the delay in answering. We have some instructions on Docker usage here: https://github.com/weserv/images/blob/4.x/DOCKER.md

The instructions can be combined with any general tutorial on Docker or running Docker images on GCP, for example: https://cloud.google.com/kubernetes-engine/docs/tutorials/hello-app

The application self is pretty light, and memory-usage can be reduced if you disable proxy_cache, but this will impact performance a bit.

The reason we block Google is explained in: https://github.com/weserv/images/issues/165
Short summary: If we don't block Google, the free service cannot exist because of all the DMCA-complaints we'd need to process. Copyright laws prevent a truly free internet for all.

Please let us know if there is specific information missing in our source code. Also, if there's a good addition to the help files; feel free to share, so we can update them. We don't use the Docker images for the free service, but they should work fine. We do run from the exact same source code (and config) as published in this repo.

**Comment by mohanrao** [comment]
Hi @andrieslouw Thanks for the detailed response.

When i am trying to run on my local as container. 

I am getting forbidden 403 as response

And also below are the logs

```
2019-04-16 07:35:03,510 CRIT Set uid to user 0
2019-04-16 07:35:03,550 INFO RPC interface 'supervisor' initialized
2019-04-16 07:35:03,551 CRIT Server 'unix_http_server' running without any HTTP authentication checking
2019-04-16 07:35:03,551 INFO supervisord started with pid 1
2019-04-16 07:35:04,560 INFO spawned: 'redis' with pid 8
2019-04-16 07:35:04,583 INFO spawned: 'nginx' with pid 9
2019-04-16 07:35:05,597 INFO success: redis entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2019-04-16 07:35:05,597 INFO success: nginx entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)


```

**Comment by mohanrao** [comment]
After reruning one more time i am getting the below logs of docker.

```
2019-04-16 23:36:54,006 CRIT Set uid to user 0
2019-04-16 23:36:54,028 INFO RPC interface 'supervisor' initialized
2019-04-16 23:36:54,028 CRIT Server 'unix_http_server' running without any HTTP authentication checking
2019-04-16 23:36:54,028 INFO supervisord started with pid 1
2019-04-16 23:36:55,033 INFO spawned: 'redis' with pid 9
2019-04-16 23:36:55,040 INFO spawned: 'nginx' with pid 10
2019-04-16 23:36:55,058 INFO exited: nginx (exit status 1; not expected)
2019-04-16 23:36:56,064 INFO success: redis entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2019-04-16 23:36:56,070 INFO spawned: 'nginx' with pid 14
2019-04-16 23:36:56,081 INFO exited: nginx (exit status 1; not expected)
2019-04-16 23:36:58,095 INFO spawned: 'nginx' with pid 15
2019-04-16 23:36:58,108 INFO exited: nginx (exit status 1; not expected)
2019-04-16 23:37:02,094 INFO spawned: 'nginx' with pid 16
2019-04-16 23:37:02,106 INFO exited: nginx (exit status 1; not expected)
2019-04-16 23:37:03,110 INFO gave up: nginx entered FATAL state, too many start retries too quickly
```

**Comment by andrieslouw** [comment]
Could you check the error logs of NGINX?

It should be something like
```
docker exec imagesweserv nginx -t
docker logs imagesweserv
docker exec imagesweserv tail /usr/local/openresty/nginx/logs/nginx-error.log
```

@kleisauke : do we redirect the NGINX logs to STDERR/STDOUT? Like in https://docs.docker.com/config/containers/logging/ ?

**Comment by mohanrao** [comment]
@andrieslouw  also in the documentation docker run is specified two times. I ran only once. 

Also does the below exec step is required ? 
```
docker exec imagesweserv make dev
```

**Comment by mohanrao** [comment]
When i ran the commands to check the logs of ngnix. I got the output as below

```
Mohanas-MBP:~ mohanarao$ docker exec imagesweserv nginx -t
nginx: [error] [lua] init_by_lua:10: Failed to disable libvips operations cache init_by_lua:3: module 'vips' not found:
	no field package.preload['vips']
	no file '/var/www/imagesweserv/app/vips.lua'
	no file '/var/www/imagesweserv/src/vips.lua'
	no file '/usr/local/openresty/site/lualib/vips.ljbc'
	no file '/usr/local/openresty/site/lualib/vips/init.ljbc'
	no file '/usr/local/openresty/lualib/vips.ljbc'
	no file '/usr/local/openresty/lualib/vips/init.ljbc'
	no file '/usr/local/openresty/site/lualib/vips.lua'
	no file '/usr/local/openresty/site/lualib/vips/init.lua'
	no file '/usr/local/openresty/lualib/vips.lua'
	no file '/usr/local/openresty/lualib/vips/init.lua'
	no file './vips.lua'
	no file '/usr/local/openresty/luajit/share/luajit-2.1.0-beta3/vips.lua'
	no file '/usr/local/share/lua/5.1/vips.lua'
	no file '/usr/local/share/lua/5.1/vips/init.lua'
	no file '/usr/local/openresty/luajit/share/lua/5.1/vips.lua'
	no file '/usr/local/openresty/luajit/share/lua/5.1/vips/init.lua'
	no file '/usr/local/openresty/site/lualib/vips.so'
	no file '/usr/local/openresty/lualib/vips.so'
	no file './vips.so'
	no file '/usr/local/lib/lua/5.1/vips.so'
	no file '/usr/local/openresty/luajit/lib/lua/5.1/vips.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
nginx: the configuration file /usr/local/openresty/nginx/conf/nginx.conf syntax is ok
nginx: configuration file /usr/local/openresty/nginx/conf/nginx.conf test is successful
Mohanas-MBP:~ mohanarao$ docker logs imagesweserv
2019-04-17 00:02:06,954 CRIT Set uid to user 0
2019-04-17 00:02:06,977 INFO RPC interface 'supervisor' initialized
2019-04-17 00:02:06,977 CRIT Server 'unix_http_server' running without any HTTP authentication checking
2019-04-17 00:02:06,977 INFO supervisord started with pid 1
2019-04-17 00:02:07,987 INFO spawned: 'redis' with pid 9
2019-04-17 00:02:08,003 INFO spawned: 'nginx' with pid 13
2019-04-17 00:02:09,025 INFO success: redis entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2019-04-17 00:02:09,025 INFO success: nginx entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
Mohanas-MBP:~ mohanarao$ docker exec imagesweserv tail /usr/local/openresty/nginx/logs/nginx-error.log
Mohanas-MBP:~ mohanarao$ 

```

**Comment by andrieslouw** [comment]
Yes, the make dev is required, according to the logs, vips is not installed now. It should be installed by make dev.

**Comment by mohanrao** [comment]
@andrieslouw i ran that later. Still i see 403 forbidden

```
Mohanas-MBP:~ mohanarao$ docker exec imagesweserv nginx -t
nginx: the configuration file /usr/local/openresty/nginx/conf/nginx.conf syntax is ok
nginx: configuration file /usr/local/openresty/nginx/conf/nginx.conf test is successful
Mohanas-MBP:~ mohanarao$ docker logs imagesweserv
2019-04-17 00:02:06,954 CRIT Set uid to user 0
2019-04-17 00:02:06,977 INFO RPC interface 'supervisor' initialized
2019-04-17 00:02:06,977 CRIT Server 'unix_http_server' running without any HTTP authentication checking
2019-04-17 00:02:06,977 INFO supervisord started with pid 1
2019-04-17 00:02:07,987 INFO spawned: 'redis' with pid 9
2019-04-17 00:02:08,003 INFO spawned: 'nginx' with pid 13
2019-04-17 00:02:09,025 INFO success: redis entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2019-04-17 00:02:09,025 INFO success: nginx entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
Mohanas-MBP:~ mohanarao$ docker exec imagesweserv tail /usr/local/openresty/nginx/logs/nginx-error.log
Mohanas-MBP:~ mohanarao$ 

```

Also if we need to run the make dev . Is there a reason why it is not part of docker build ?

**Comment by andrieslouw** [comment]
I think the 403 may be due to https://github.com/weserv/images/blob/73c3f4b5e161eac9d2c90550a3d489f586504305/config/nginx/conf.d/imagesweserv.conf#L40-L41

For testing we only allow connections from localhost, remove those lines, and things should work.

**Comment by andrieslouw** [comment]
How do you access the server? It should listen to any host on port 80. But images.weserv.local is just internal. Strange that the 403 is not showing up in any logs. @kleisauke : could you assist as to the default location of the logfiles?

**Comment by mohanrao** [comment]
Hi @andrieslouw  and @kleisauke  any update on this?

**Comment by kleisauke** [comment]
It seems that the internal server (`images.weserv.local`) was chosen by nginx as the default server instead of the catch-all server. This resulted in the 403 forbidden errors.

Commit https://github.com/weserv/images/commit/4ae9a2269aa4ec0c32037538c1cf7d35c5ab62c0 should fix this. In that commit, I also fixed the error/access log path locations to be forwarded to the Docker log collector and added some useful commands in [`DOCKER.md`](https://github.com/weserv/images/blob/4.x/DOCKER.md) to view the error/access logs.

I've successfully tested these changes in a local container using the Google Cloud Shell.

**Comment by mohanrao** [comment]
Thanks @kleisauke .

- Why can't we have a official docker image on docker hub

- Also can we have below step as docker build itself if it is required one 
```
docker exec imagesweserv make dev
```

**Comment by kleisauke** [comment]
@mohanrao I've just created #180 to improve the Docker image and deployment.

**Comment by kleisauke** [comment]
Update:
Almost all tasks are now resolved in #180. It will now be easier to use the Docker image on the Google Cloud Platform.

I will soon publish the revamped Docker image on [Docker Hub](https://hub.docker.com/u/weserv).

**Comment by andrieslouw** [comment]
Will continue in #204 as that is the only task remaining AFAIK.
Closing this issue. Let us know if there is anything we can do to assist.
